pytorch_lightning/trainer/training_loop.py

```diff
@@ -433,6 +433,13 @@ def run_training_batch(self, batch, batch_idx):
 
             # call training_step once per optimizer
             for opt_idx, optimizer in enumerate(self.optimizers):
+                # make sure only the gradients of the current optimizer's paramaters are calculated 
+                # in the training step to prevent dangling gradients in multiple-optimizer setup.
+                for param in self.get_model().parameters():
+                    param.requires_grad = False
+                for group in optimizer.param_groups:
+                    for param in group['params']:
+                        param.requires_grad = True
```

Comment:
should all params of all optimizers be set to True after the step in order to restore the original state? I can't think of a concrete example but wonder if this may lead to very hard to find bugs.

Reply:
yeah, i think this should be a freeze, unfreeze type of op. agreed

Reply:
but we really should keep exactly the same state as before. So, blanket freeze unfreeze is bad. we should freeze unfrozen params and unfreeze only those after (keep a dict of indices when freezing, for example)

Reply:
I also agree, a user may want to inspect gradients sometime in training. Helpers like save_freeze_state() and restore_freeze_state() would definitely help. Also, we only need this logic in multiple optimizer training setup so we can add an "if len(self.optimizers) > 1" to the top so that we can avoid redundantly freezing and unfreezing all parameters at every batch which takes ~0.5% of the time.

Reply:
@aybberk @williamFalcon - this change perfectly solves any dangling grads issue, but seems to work much slower than the suggested fix at #666 

The change in #666 resets the grads ones - thus communicates with the GPU once at every training step. 
This solution passes on each layer to set require_grads, and if the model runs on a GPU, the GPU will be addressed many times in a row.

From my experiments with this approach, it slows the runtime of the training by a significant factor. Have you benchmarked it yourself? What were your results?

Reply:
> @aybberk @williamFalcon - this change perfectly solves any dangling grads issue, but seems to work much slower than the suggested fix at #666
> 
> The change in #666 resets the grads ones - thus communicates with the GPU once at every training step.
> This solution passes on each layer to set require_grads, and if the model runs on a GPU, the GPU will be addressed many times in a row.
> 
> From my experiments with this approach, it slows the runtime of the training by a significant factor. Have you benchmarked it yourself? What were your results?

Benchmarks are done in my local pc with gtx1060, and shown a negligible effect(~0.5%) on performance. How was your results? I think multiple sequential GPU communication is not a bottleneck there since when you call a zero_grad() on an optimizer or model, PyTorch does the exact same thing in the lower level, to the best of my information.